### PR TITLE
Destinations V2: clean up AirbyteType code

### LINKED ISSUE #28054: 1s1t: cleanup+refactoring in AirbyteType/Utils
do @jbfbell 's comments

* this entire review https://github.com/airbytehq/airbyte/pull/27936#pullrequestreview-1515105919
* https://github.com/airbytehq/airbyte/pull/27936#discussion_r1253589757

### PR DESCRIPTION
**Pull request by cynthiaxyin**
closes #28054

- Addresses feedback on #27936 for the AirbyteType code area
- Simplifies the protocol type ordering and exclusion logic
- Other code cleanup and hygiene

Recommended reading order:
- Work is broken down and labeled commit by commit!

### CODE REVIEW

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java [hunk 1]
```diff
@@ -4,153 +4,87 @@
...
-          for (final AirbyteProtocolType excludedType : excludedTypes) {
-            if (typePresenceMap.getOrDefault(excludedType, false)) {
-              foundExcludedTypes = true;
-              break;
-            }
-          }
-          if (!foundExcludedTypes) {
-            return protocolType;
+  // Picks which type in a Union takes precedence
+  public static AirbyteType chooseUnionType(final Union u) {
```
Comment (cynthiaxyin) [inline airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java:73]: much simpler and easier to read than before :)

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`f166ec5a72`](https://github.com/airbytehq/airbyte/pull/28430/commits/f166ec5a727003d01451ee3c4cd6b3dd04f923e3)) - ✅

⏲️  Total pipeline duration: 23mn42s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689791517/f166ec5a727003d01451ee3c4cd6b3dd04f923e3/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`f166ec5a72`](https://github.com/airbytehq/airbyte/pull/28430/commits/f166ec5a727003d01451ee3c4cd6b3dd04f923e3)) - ❌

⏲️  Total pipeline duration: 39mn15s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689791517/f166ec5a727003d01451ee3c4cd6b3dd04f923e3/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`19342023f3`](https://github.com/airbytehq/airbyte/pull/28430/commits/19342023f30e09509e6de6eb2c6b61e21974eda8)) - ❌

⏲️  Total pipeline duration: 18mn09s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689795397/19342023f30e09509e6de6eb2c6b61e21974eda8/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`19342023f3`](https://github.com/airbytehq/airbyte/pull/28430/commits/19342023f30e09509e6de6eb2c6b61e21974eda8)) - ✅

⏲️  Total pipeline duration: 27mn58s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689795397/19342023f30e09509e6de6eb2c6b61e21974eda8/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

File: airbyte-integrations/bases/base-typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeTest.java [hunk 2]
```diff
@@ -417,64 +421,55 @@ public void testInvalid() {
...
+    unionToType.put(new Union(ImmutableList.of(s, a)), a);
+    unionToType.put(new Union(ImmutableList.of(NUMBER, a)), a);
+    unionToType.put(new Union(ImmutableList.of(INTEGER, s)), s);
+    unionToType.put(new Union(ImmutableList.of(NUMBER, DATE, BOOLEAN)), DATE);
+    unionToType.put(new Union(ImmutableList.of(INTEGER, BOOLEAN, NUMBER)), NUMBER);
+    unionToType.put(new Union(ImmutableList.of(BOOLEAN, INTEGER)), INTEGER);
 
-    o = new OneOf(ImmutableList.of(AirbyteProtocolType.BOOLEAN, AirbyteProtocolType.NUMBER, AirbyteProtocolType.STRING));
-    assertEquals(AirbyteProtocolType.STRING, AirbyteTypeUtils.chooseOneOfType(o));
+    unionToType.forEach((u, t) -> assertEquals(t, chooseUnionType(u)));
```
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeTest.java:446]: nit: might want to do something with `assertAll` here

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java [hunk 3]
```diff
@@ -4,153 +4,85 @@
...
-          return true;
-        }
-      }
-    }
-    return false;
+    return node.equals(TextNode.valueOf(value));
   }
 
+  // Extracts the appropriate protocol type from the representative JSON
   protected static AirbyteType getAirbyteProtocolType(final JsonNode node) {
```
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java:32]: does this always return `AirbyteProtocolType`? would be nice to tighten the return type here

and then we can make this a static method inside AirbyteProtocolType

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java [hunk 4]
```diff
@@ -4,153 +4,85 @@
...
-    // looping through the options only once for efficiency
-    for (final AirbyteType option : options) {
-      if (option instanceof final Array a) {
-        foundArrayType = a;
-      } else if (option instanceof final Struct s) {
-        foundStructType = s;
-      } else if (option instanceof final AirbyteProtocolType p) {
-        typePresenceMap.put(p, true);
+  // Picks which type in a Union takes precedence
+  public static AirbyteType chooseUnionType(final Union u) {
```
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java:73]: imo we should move this to be a method inside the Union class

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteType.java [hunk 5]
```diff
@@ -107,17 +83,57 @@ private static Array getArray(final JsonNode schema) {
     }
   }
 
+  private static AirbyteType fromArrayJsonSchema(final JsonNode schema, final JsonNode array) {
```
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteType.java:86]: I think we're at a point where all these inner classes should just get their own file... feels like having them all in one file is doing more harm than good

**Review by edgao — APPROVED** [review]
had a few comments but nothing blocking. I'd weakly like to completely kill the AirbyteTypeUtils class and have those two methods both live inside AirbyteType.java, but only very weakly.

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java [hunk 6]
```diff
@@ -4,153 +4,87 @@
...
+      return BOOLEAN;
+    } else if (nodeMatches(propertyType, "integer")) {
+      return INTEGER;
+    } else if (nodeMatches(propertyType, "number")) {
+      return nodeMatches(airbyteType, "integer") ? INTEGER : NUMBER;
+    } else if (nodeMatches(propertyType, "string")) {
+      if (nodeMatches(format, "date")) {
+        return DATE;
+      } else if (nodeMatches(format, "time")) {
+        if (nodeMatches(airbyteType, "timestamp_without_timezone")) {
```
[the remark this answers is not in the document]
Comment (cynthiaxyin) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java:53]: Yes, I removed that commit from this change. Will rebase to pick that up.

File: airbyte-integrations/bases/base-typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeTest.java (hunk 2 above)
Comment (cynthiaxyin) [reply airbyte-integrations/bases/base-typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeTest.java]: I wrapped the line in an assertAll but lmk if that's not what you had in mind.

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`65f5547ca7`](https://github.com/airbytehq/airbyte/pull/28430/commits/65f5547ca7c36967e5f95591e018041e4267423b)) - ✅

⏲️  Total pipeline duration: 17mn06s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689815384/65f5547ca7c36967e5f95591e018041e4267423b/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`65f5547ca7`](https://github.com/airbytehq/airbyte/pull/28430/commits/65f5547ca7c36967e5f95591e018041e4267423b)) - ✅

⏲️  Total pipeline duration: 26mn49s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689815384/65f5547ca7c36967e5f95591e018041e4267423b/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`6e82abad12`](https://github.com/airbytehq/airbyte/pull/28430/commits/6e82abad12145be58548970092a42ca209746097)) - ✅

⏲️  Total pipeline duration: 18mn10s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689818104/6e82abad12145be58548970092a42ca209746097/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`6e82abad12`](https://github.com/airbytehq/airbyte/pull/28430/commits/6e82abad12145be58548970092a42ca209746097)) - ✅

⏲️  Total pipeline duration: 26mn06s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689818104/6e82abad12145be58548970092a42ca209746097/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`ae42899a1b`](https://github.com/airbytehq/airbyte/pull/28430/commits/ae42899a1b14ceeede806e1d39b363babc7ff5fe)) - ✅

⏲️  Total pipeline duration: 25mn41s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689880293/ae42899a1b14ceeede806e1d39b363babc7ff5fe/destination-bigquery/1.5.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`ae42899a1b`](https://github.com/airbytehq/airbyte/pull/28430/commits/ae42899a1b14ceeede806e1d39b363babc7ff5fe)) - ✅

⏲️  Total pipeline duration: 17mn40s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689880293/ae42899a1b14ceeede806e1d39b363babc7ff5fe/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`ae42899a1b`](https://github.com/airbytehq/airbyte/pull/28430/commits/ae42899a1b14ceeede806e1d39b363babc7ff5fe)) - ✅

⏲️  Total pipeline duration: 24mn44s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689885245/ae42899a1b14ceeede806e1d39b363babc7ff5fe/destination-bigquery/1.5.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`ae42899a1b`](https://github.com/airbytehq/airbyte/pull/28430/commits/ae42899a1b14ceeede806e1d39b363babc7ff5fe)) - ✅

⏲️  Total pipeline duration: 25.97s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/cynthia_airbyte_type_cleanup/1689885245/ae42899a1b14ceeede806e1d39b363babc7ff5fe/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Pull request merged**